python/cudf/cudf/tests/test_groupby.py

```diff
@@ -2972,3 +2973,73 @@ def test_groupby_dtypes(groups):
     pdf = df.to_pandas()
 
     assert_eq(pdf.groupby(groups).dtypes, df.groupby(groups).dtypes)
+
+
+class TestSample:
+    @pytest.fixture(
+        params=[
+            ["a", "a", "b", "b", "c", "c", "c", "d", "d", "d", "d", "d"],
+            [1, 1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 4],
+        ],
+        ids=["str-group", "int-group"],
+    )
+    def df(self, request):
+        return cudf.DataFrame(
+            {"a": request.param, "b": request.param, "v": request.param}
+        )
+
+    @pytest.fixture(params=["a", ["a", "b"]], ids=["single-col", "two-col"])
+    def by(self, request):
+        return request.param
+
+    def expected(self, df, *, n=None, frac=None):
+        value_counts = collections.Counter(df.a.values_host)
+        if n is not None:
+            values = list(
+                itertools.chain.from_iterable(
+                    itertools.repeat(v, n) for v in value_counts.keys()
+                )
+            )
+        elif frac is not None:
+            values = list(
+                itertools.chain.from_iterable(
+                    itertools.repeat(v, round(count * frac))
+                    for v, count in value_counts.items()
+                )
+            )
+        else:
+            raise AssertionError("Invalid usage")
+        values = cudf.Series(sorted(values), dtype=df.a.dtype)
+        return cudf.DataFrame({"a": values, "b": values, "v": values})
+
+    @pytest.mark.parametrize("n", [None, 0, 1, 2])
+    def test_constant_n_no_replace(self, df, by, n):
+        result = df.groupby(by).sample(n=n).sort_values("a")
+        n = 1 if n is None else n
+        assert_eq(self.expected(df, n=n), result.reset_index(drop=True))
+
+    def test_constant_n_no_replace_too_large_raises(self, df):
+        with pytest.raises(ValueError):
+            df.groupby("a").sample(n=3)
+
+    @pytest.mark.parametrize("n", [1, 2, 3])
+    def test_constant_n_replace(self, df, by, n):
+        result = df.groupby(by).sample(n=n, replace=True).sort_values("a")
+        assert_eq(self.expected(df, n=n), result.reset_index(drop=True))
+
+    def test_invalid_arguments(self, df):
+        with pytest.raises(ValueError):
+            df.groupby("a").sample(n=1, frac=0.1)
+
+    def test_not_implemented_arguments(self, df):
+        with pytest.raises(NotImplementedError):
+            # These are valid weights, but we don't implemented this yet.
+            df.groupby("a").sample(n=1, weights=[1 / len(df)] * len(df))
+
+    @pytest.mark.parametrize("frac", [0, 1 / 3, 1 / 2, 2 / 3, 1])
+    @pytest.mark.parametrize("replace", [False, True])
+    def test_fraction_rounding(self, df, by, frac, replace):
+        result = (
+            df.groupby(by).sample(frac=frac, replace=replace).sort_values("a")
+        )
+        assert_eq(self.expected(df, frac=frac), result.reset_index(drop=True))
```

Comment:
TODO, these tests don't check for statistical correctness that the sampling in each group is uniform.

Reply:
tl;dr: this works.

I did some statistical tests using this test dataframe:
```
test_df = cudf.DataFrame({"a": [1, 1, 1, 1, 1, 2, 2, 2, 2, 2], "b": [0, 1, 2, 3, 4, 0, 1, 2, 3, 4]}
```
So if I groupby "a" and `sample(n=4, replace=False)` I should draw length four subsequences from all length five permutations for both groups. For this setup there are 120 different results. So the expected probability of obtaining any given sample is 1/120.

Doing this 50_000 times, I observe the empirical distribution of the sample counts, and use `scipy.stats.normaltest` to check for normality. I can't reject at the 5% significance level (p=0.7) and so the null hypothesis that the data are from a normal distribution is reasonable.

I also use a nonparametric Kolmogorov-Smirnov test to compare to sample drawn using `np.random.choice([0, 1, 2, 3, 4], size=4, replace=False)` and again cannot reject the null hypothesis that the two datasets were drawn from the same distribution.

For sampling with replacement, I choose `n=3` (125 possible outcomes) and do the same thing again. With 50_000 samples I cannot reject the null hypothesis that the data are normal, nor can I reject the null hypothesis that the data are from the same underlying distribution as the "naive" numpy implementation.

So, I think this is good.